design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,542 @@
+{
+    void GetError(xlang_result* error);
+    void GetMessage(xlang_string* message);
+    void GetLanguageError(xlang_string* error);
+    void GetExecutionTrace(IXlangErrorExecutionTrace* executionTrace);
+    void GetLanguageIdentifier(xlang_string* identifier);
+    void GetLanguageInformation(IUnknown* languageInformation);
+    void GetLanguageInformationAsString(xlang_string* languageInformation);
+    void GetPropagatedError(IXlangErrorInfo** propagatedError);
+    void PropagateError(
+        xlang_string projectionIdentifier,
+        xlang_string languageError,
+        IXlangErrorExecutionTrace* executionTrace,
+        IUnknown* languageInformation);
+}
+```
+
+Most of the functions provided in this interface are self-explanatory. But there are a few which
+should be called out. There are 2 variants of the function used to get the additional language
+specific information (`GetLanguageInformation` and `GetLanguageInformationAsString`).
+`GetLanguageInformation` should be used in scenarios where the projection itself wants to retrieve
+an object they stored. After retrieving the object, the projection should verify it is their object
+by querying for some custom interface they implement on it.
+`GetLanguageInformationAsString` should be used in scenarios where a projection wants to retrieve the
+additional information stored to expose it as part of the language function used to log errors.
+The function will provide the information as a string if the object stored implements the
+`IXlangErrorExtendedInfo` interface. If it doesn't or no object was stored, it will just provide an
+empty string.
+
+Another function to call out is the `GetPropagatedError` function. This function will allow
+to traverse the diagnostic information provided by the projections which the error propagated through.
+For each error info retrieved from it, the error and message functions will provide the same value as the
+initial error info. The remaining functions will provide the respective information provided the projection.
+It was decided to keep all these functions as part of the same interface rather than splitting them out to
+avoid having projections query for multiple interfaces to get details from the error info.
+
+## Error free functions
+
+Some languages have a concept of error free functions (i.e. noexcept in C++). These functions
+will not be projected as error free because errors can occur while translating the call from the
+other language and while translating the result back (i.e. during activation of the class for the
+function). But if there was any error during the execution of the function itself, the language
+itself may take actions based on what it does for errors in error free functions (i.e. fail fast for
+noexcept).
+
+Even though none of the projected functions will be error free, the PAL can consist of error free
+functions. These functions can be distinguished by the return parameter being void instead of
```

Comment:
per discussion - the ABI must never change due to introduction (or removal) of the noexcept attribute.  only the projection may choose to optimize it's codegen based on noexcept - both to remove an if check and to remove catch block at the call site.  an ABI C header could choose to omit [[nodiscard]] on [noexcept] methods, however

Reply:
Addressed.